File: codechef/WICP2002/02_chef_and_jail/chef_and_jail.py
#!/usr/bin/env python
# https://www.codechef.com/WICP2002/problems/CHJAIL

from sys import stdin

def read():
    input_data = stdin.read().split()
    for token in input_data:
        yield token

def tokens(g, n):
    t = []
    for _ in range(n):
        t.append(next(g))
    return t


def escapable(jail, m, n, target):

    def next_rooms(room):
        return filter(lambda r: r[0] <= target[0] and r[1] <= target[1],
                      [(room[0], room[1]+1), (room[0]+1, room[1]), (room[0]+1, room[1]+1)])

    def aux(room, power):
        if room == target:
            return power >= 0
        if power > 0:
            for r in next_rooms(room):
                if aux(r, power - jail[r[0]][r[1]]):
                    return True
        return False

    return aux((0, 0), jail[0][0])


reader = read()
for _ in range(int(next(reader))):
    m, n, x, y = map(int, tokens(reader, 4))
    jail = []
    for row in range(m):
        jail.append(list(map(int, tokens(reader, n))))
    if escapable(jail, m, n, (y-1, x-1)):
        print('Escape')
    else:
        print('Died')

File: codechef/WICP2002/02_chef_and_jail/chef_and_jail.cc
#include <bits/stdc++.h>

using namespace std;

int m, n, x, y;
vector<vector<int>> jail;

bool aux(int i, int j, int p) {
  if (i == y && j == x) return p >= 0;
  int dy[] = {1, 0, 1};
  int dx[] = {0, 1, 1};
  if (p > 0)
    for (int k = 0; k < 3; k++) {
      int ii = i+dy[k];
      int jj = j+dx[k];
      if (ii > y || jj > x) continue;
      if (aux(ii, jj, p - jail[ii][jj])) return true;
    }
  return false;
}

bool escapable() {
  return aux(0, 0, jail[0][0]);
}

int main() {
  int t;
  cin >> t;
  while (t--) {
    cin >> m >> n >> x >> y;
    x--, y--;
    jail = vector<vector<int>>(m, vector<int>(n));
    for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++)
        cin >> jail[i][j];
    if (escapable()) cout << "Escape\n";
    else cout << "Died\n";
  }
}
